File: sdks/rust_sdk/src/ack_integration.rs
use crate::activity_buffer::PersistentActivityBuffer;
use crate::acks::{build_ack_envelope, map_error_to_error_code};
use crate::clients::RouterClient;
use crate::constants::*;
use crate::envelope::EnvelopeData;
use crate::persistence::PersistentActivityRecord;
use crate::{Error, Result};
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::RwLock;

/// Result of sending a message with ACK tracking
#[derive(Debug, Clone)]
pub struct SendResult {
    pub success: bool,
    pub message_id: String,
    pub accepted: bool,
    pub reason: String,
    pub activity_record: Option<PersistentActivityRecord>,
}

/// ACK lifecycle manager for message tracking and automatic acknowledgments
pub struct AckLifecycleManager {
    router: Arc<RwLock<RouterClient>>,
    buffer: Arc<PersistentActivityBuffer>,
    agent_id: String,
    auto_ack: bool,
    ack_timeout: Duration,
}

impl AckLifecycleManager {
    pub fn new(
        router: RouterClient,
        buffer: PersistentActivityBuffer,
        agent_id: String,
        auto_ack: bool,
        ack_timeout_seconds: u64,
    ) -> Self {
        Self {
            router: Arc::new(RwLock::new(router)),
            buffer: Arc::new(buffer),
            agent_id,
            auto_ack,
            ack_timeout: Duration::from_secs(ack_timeout_seconds),
        }
    }

    /// Send a message with automatic ACK lifecycle handling
    pub async fn send_message_with_ack(&self, envelope: EnvelopeData) -> Result<SendResult> {
        let message_id = envelope.message_id.clone();

        // Record outgoing message
        let activity_record = self.buffer.record_outgoing(
            serde_json::to_value(&envelope).map_err(Error::Serialization)?
        )?;

        match self._send_message_internal(&envelope).await {
            Ok((accepted, reason)) => {
                // Generate automatic ACK based on result
                if self.auto_ack {
                    let ack_stage = if accepted {
                        ack_stage::FULFILLED
                    } else {
                        ack_stage::REJECTED
                    };
                    
                    let error_code = if accepted {
                        error_code::ERROR_CODE_UNSPECIFIED
                    } else {
                        error_code::VALIDATION_ERROR
                    };

                    // Update activity record via ACK processing
                    let ack_data = serde_json::json!({
                        "ack_for_message_id": message_id,
                        "ack_stage": ack_stage,
                        "error_code": error_code,
                        "note": reason
                    });

                    self.buffer.ack(&ack_data)?;
                    self.buffer.flush()?;
                }

                Ok(SendResult {
                    success: true,
                    message_id,
                    accepted,
                    reason,
                    activity_record: Some(activity_record),
                })
            }
            Err(e) => {
                // Handle send failure
                let error_code = map_error_to_error_code(&e);
                let ack_data = serde_json::json!({
                    "ack_for_message_id": message_id,
                    "ack_stage": ack_stage::FAILED,
                    "error_code": error_code,
                    "note": e.to_string()
                });

                self.buffer.ack(&ack_data)?;
                self.buffer.flush()?;

                Ok(SendResult {
                    success: false,
                    message_id,
                    accepted: false,
                    reason: e.to_string(),
                    activity_record: Some(activity_record),
                })
            }
        }
    }

    async fn _send_message_internal(&self, envelope: &EnvelopeData) -> Result<(bool, String)> {
        let mut router = self.router.write().await;
        
        // Convert our envelope to the format expected by RouterClient
        router.send_message(envelope).await?;
        
        // For now, assume success since RouterClient.send_message doesn't return acceptance info
        // In a real implementation, this would parse the actual router response
        Ok((true, "Message sent successfully".to_string()))
    }

    /// Process an incoming ACK message
    pub fn process_incoming_ack(&self, envelope: &EnvelopeData) -> Result<Option<PersistentActivityRecord>> {
        if envelope.message_type != message_type::ACKNOWLEDGEMENT {
            return Ok(None);
        }

        // Parse ACK from envelope payload
        let ack_data: Value = envelope.json_payload()
            .map_err(|e| Error::InvalidEnvelope(format!("Invalid ACK payload: {}", e)))?;

        // Update activity buffer
        match self.buffer.ack(&ack_data)? {
            Some(record) => {
                self.buffer.flush()?;
                Ok(Some(record))
            }
            None => Ok(None),
        }
    }

    /// Send an ACK message for a received message
    pub async fn send_ack(
        &self,
        ack_for_message_id: String,
        stage: i32,
        error_code: Option<i32>,
        note: Option<String>,
    ) -> Result<SendResult> {
        let ack_envelope = build_ack_envelope(
            self.agent_id.clone(),
            ack_for_message_id,
            stage,
            error_code,
            note,
            None,
            None,
        )?;

        self.send_message_with_ack(ack_envelope).await
    }

    /// Get outgoing messages that need ACK reconciliation
    pub fn get_unacked_outgoing(&self) -> Result<Vec<PersistentActivityRecord>> {
        let unacked = self.buffer.unacked()?;
        let outgoing: Vec<PersistentActivityRecord> = unacked
            .into_iter()
            .filter(|record| record.direction == "out")
            .collect();
        Ok(outgoing)
    }

    /// Get incoming messages that still need ACKs to be sent
    pub fn get_pending_acks(&self) -> Result<Vec<PersistentActivityRecord>> {
        let unacked = self.buffer.unacked()?;
        let pending: Vec<PersistentActivityRecord> = unacked
            .into_iter()
            .filter(|record| {
                record.direction == "in" && 
                matches!(record.ack_stage, 
                    ack_stage::ACK_STAGE_UNSPECIFIED | 
                    ack_stage::RECEIVED
                )
            })
            .collect();
        Ok(pending)
    }

    /// Find messages that may need retry or follow-up ACKs
    pub fn reconcile_acks(&self) -> Result<Vec<PersistentActivityRecord>> {
        let now = chrono::Utc::now().timestamp_millis();
        let timeout_ms = self.ack_timeout.as_millis() as i64;

        let unacked = self.buffer.unacked()?;
        let stale: Vec<PersistentActivityRecord> = unacked
            .into_iter()
            .filter(|record| {
                let age_ms = now - record.ts_ms;
                age_ms > timeout_ms
            })
            .collect();

        Ok(stale)
    }

    /// Automatically send RECEIVED ACK for incoming message
    pub async fn auto_ack_received(&self, envelope: &EnvelopeData) -> Result<SendResult> {
        // Record incoming message
        let envelope_json = serde_json::to_value(envelope).map_err(Error::Serialization)?;
        self.buffer.record_incoming(envelope_json)?;

        // Send RECEIVED ACK
        self.send_ack(
            envelope.message_id.clone(),
            ack_stage::RECEIVED,
            None,
            None,
        ).await
    }

    /// Send READ ACK to indicate message has been processed
    pub async fn auto_ack_read(&self, message_id: String) -> Result<SendResult> {
        self.send_ack(message_id, ack_stage::READ, None, None).await
    }

    /// Send FULFILLED ACK to indicate successful processing
    pub async fn auto_ack_fulfilled(&self, message_id: String, note: Option<String>) -> Result<SendResult> {
        self.send_ack(message_id, ack_stage::FULFILLED, None, note).await
    }

    /// Send REJECTED ACK to indicate processing rejection
    pub async fn auto_ack_rejected(
        &self,
        message_id: String,
        reason: String,
        error_code: Option<i32>,
    ) -> Result<SendResult> {
        let code = error_code.unwrap_or(error_code::VALIDATION_ERROR);
        self.send_ack(message_id, ack_stage::REJECTED, Some(code), Some(reason)).await
    }

    /// Send FAILED ACK to indicate processing failure
    pub async fn auto_ack_failed(&self, message_id: String, error: &Error) -> Result<SendResult> {
        let error_code = map_error_to_error_code(error);
        self.send_ack(
            message_id,
            ack_stage::FAILED,
            Some(error_code),
            Some(error.to_string()),
        ).await
    }
}

/// High-level message processor with automatic ACK handling
pub struct MessageProcessor {
    ack_manager: AckLifecycleManager,
    message_handlers: HashMap<i32, Box<dyn Fn(&EnvelopeData) -> Result<Value> + Send + Sync>>,
    default_handler: Option<Box<dyn Fn(&EnvelopeData) -> Result<Value> + Send + Sync>>,
}

impl MessageProcessor {
    pub fn new(ack_manager: AckLifecycleManager) -> Self {
        Self {
            ack_manager,
            message_handlers: HashMap::new(),
            default_handler: None,
        }
    }

    /// Register a handler for a specific message type
    pub fn register_handler<F>(&mut self, message_type: i32, handler: F)
    where
        F: Fn(&EnvelopeData) -> Result<Value> + Send + Sync + 'static,
    {
        self.message_handlers.insert(message_type, Box::new(handler));
    }

    /// Set a default handler for unregistered message types
    pub fn set_default_handler<F>(&mut self, handler: F)
    where
        F: Fn(&EnvelopeData) -> Result<Value> + Send + Sync + 'static,
    {
        self.default_handler = Some(Box::new(handler));
    }

    /// Process an incoming message with automatic ACK handling
    pub async fn process_message(&self, envelope: &EnvelopeData) -> Result<SendResult> {
        let message_type = envelope.message_type;
        let message_id = envelope.message_id.clone();

        // Send RECEIVED ACK immediately
        self.ack_manager.auto_ack_received(envelope).await?;

        // Handle ACK messages specially
        if message_type == message_type::ACKNOWLEDGEMENT {
            self.ack_manager.process_incoming_ack(envelope)?;
            return Ok(SendResult {
                success: true,
                message_id,
                accepted: true,
                reason: "ACK processed".to_string(),
                activity_record: None,
            });
        }

        // Send READ ACK before processing
        self.ack_manager.auto_ack_read(message_id.clone()).await?;

        // Find and execute handler
        let handler_result = if let Some(handler) = self.message_handlers.get(&message_type) {
            handler(envelope)
        } else if let Some(default_handler) = &self.default_handler {
            default_handler(envelope)
        } else {
            return self.ack_manager.auto_ack_rejected(
                message_id,
                format!("No handler for message type {}", message_type),
                Some(error_code::UNSUPPORTED_MESSAGE_TYPE),
            ).await;
        };

        match handler_result {
            Ok(_result) => {
                // Send FULFILLED ACK on successful processing
                self.ack_manager.auto_ack_fulfilled(
                    message_id,
                    Some("Processed successfully".to_string()),
                ).await
            }
            Err(e) => {
                // Send FAILED ACK on processing error
                self.ack_manager.auto_ack_failed(message_id, &e).await
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::persistence::JsonFilePersistence;
    use tempfile::NamedTempFile;

    #[tokio::test]
    async fn test_ack_lifecycle_manager() {
        let temp_file = NamedTempFile::new().unwrap();
        let persistence = Box::new(JsonFilePersistence::new(temp_file.path()));
        let buffer = PersistentActivityBuffer::new(1000, Some(persistence)).unwrap();
        
        // Note: We can't easily test with a real RouterClient without a running server
        // This is a placeholder for the pattern
        // let router = RouterClient::new("http://localhost:50052").await.unwrap();
        // let manager = AckLifecycleManager::new(router, buffer, "test-agent".to_string(), true, 30);
    }
}

File: sdks/py_sdk/sw4rm/ack_integration.py
from __future__ import annotations

import json
import time
from typing import Any, Dict, Optional, Callable, List
from dataclasses import dataclass

from .clients.router import RouterClient
from .activity_buffer import PersistentActivityBuffer
from .persistence import PersistentActivityRecord
from .acks import build_ack_envelope
from .error_mapping import ErrorCodeMapper, DEFAULT_MAPPER
from . import constants as C


@dataclass
class SendResult:
    """Result of sending a message with ACK tracking."""
    success: bool
    message_id: str
    accepted: bool
    reason: str = ""
    activity_record: Optional[PersistentActivityRecord] = None


class ACKLifecycleManager:
    """Manages ACK lifecycle integration with router responses and activity buffer."""

    def __init__(
        self,
        router_client: RouterClient,
        activity_buffer: PersistentActivityBuffer,
        agent_id: str,
        *,
        auto_ack: bool = True,
        ack_timeout_seconds: Optional[int] = None,
        error_mapper: Optional[ErrorCodeMapper] = None
    ):
        self.router = router_client
        self.buffer = activity_buffer
        self.agent_id = agent_id
        self.auto_ack = auto_ack
        # Spec default ACK timeout is 10s; align with spec for symmetry
        self.ack_timeout_seconds = ack_timeout_seconds or 10
        self.error_mapper = error_mapper or DEFAULT_MAPPER

    def send_message_with_ack(self, envelope: Dict[str, Any]) -> SendResult:
        """Send a message and automatically handle ACK lifecycle."""
        message_id = envelope.get("message_id", "")
        
        try:
            # Record outgoing message
            activity_record = self.buffer.record_outgoing(envelope)
            
            # Send to router
            response = self.router.send_message(envelope)
            
            # Process router response
            accepted = getattr(response, 'accepted', False)
            reason = getattr(response, 'reason', '')
            
            # Generate automatic ACK based on router response
            if self.auto_ack:
                ack_stage = C.FULFILLED if accepted else C.REJECTED
                error_code = C.ERROR_CODE_UNSPECIFIED if accepted else C.VALIDATION_ERROR
                
                # Update activity record
                activity_record.ack(ack_stage, error_code, reason)
                self.buffer.flush()

            return SendResult(
                success=True,
                message_id=message_id,
                accepted=accepted,
                reason=reason,
                activity_record=activity_record
            )

        except Exception as e:
            # Handle send failure using configurable error mapper
            error_code = self.error_mapper.map_exception(e)
            
            if message_id in self.buffer._by_id:
                self.buffer._by_id[message_id].ack(C.FAILED, error_code, str(e))
                self.buffer.flush()

            return SendResult(
                success=False,
                message_id=message_id,
                accepted=False,
                reason=str(e),
                activity_record=self.buffer.get(message_id)
            )

    def process_incoming_ack(self, envelope: Dict[str, Any]) -> Optional[PersistentActivityRecord]:
        """Process an incoming ACK message and update activity records."""
        try:
            # Parse ACK from envelope payload
            if envelope.get("message_type") != C.ACKNOWLEDGEMENT:
                return None

            payload = envelope.get("payload", b"")
            if isinstance(payload, bytes):
                ack_data = json.loads(payload.decode('utf-8'))
            else:
                ack_data = payload

            # Update activity buffer
            updated_record = self.buffer.ack(ack_data)
            if updated_record:
                self.buffer.flush()
            
            return updated_record

        except (json.JSONDecodeError, KeyError) as e:
            print(f"[ACK] Failed to process incoming ACK: {e}")
            return None

    def send_ack(self, ack_for_message_id: str, stage: int, error_code: int = C.ERROR_CODE_UNSPECIFIED, note: str = "") -> SendResult:
        """Send an ACK message for a received message."""
        ack_envelope = build_ack_envelope(
            producer_id=self.agent_id,
            ack_for_message_id=ack_for_message_id,
            ack_stage=stage,
            error_code=error_code,
            note=note
        )
        
        return self.send_message_with_ack(ack_envelope)

    def get_unacked_outgoing(self) -> List[PersistentActivityRecord]:
        """Get outgoing messages that need ACK reconciliation."""
        return [
            record for record in self.buffer.unacked()
            if record.direction == "out"
        ]

    def get_pending_acks(self) -> List[PersistentActivityRecord]:
        """Get incoming messages that still need ACKs to be sent."""
        return [
            record for record in self.buffer.unacked()
            if record.direction == "in" and record.ack_stage in (C.ACK_STAGE_UNSPECIFIED, C.RECEIVED)
        ]

    def reconcile_acks(self) -> List[PersistentActivityRecord]:
        """Find messages that may need retry or follow-up ACKs."""
        now = int(time.time() * 1000)
        timeout_ms = self.ack_timeout_seconds * 1000
        
        stale_records = []
        for record in self.buffer.unacked():
            age_ms = now - record.ts_ms
            if age_ms > timeout_ms:
                stale_records.append(record)
        
        return stale_records

    def auto_ack_received(self, envelope: Dict[str, Any]) -> SendResult:
        """Automatically send RECEIVED ACK for incoming message."""
        message_id = envelope.get("message_id", "")
        
        # Record incoming message first
        self.buffer.record_incoming(envelope)
        
        # Send RECEIVED ACK
        return self.send_ack(message_id, C.RECEIVED)

    def auto_ack_read(self, message_id: str) -> SendResult:
        """Send READ ACK to indicate message has been processed."""
        return self.send_ack(message_id, C.READ)

    def auto_ack_fulfilled(self, message_id: str, note: str = "") -> SendResult:
        """Send FULFILLED ACK to indicate successful processing."""
        return self.send_ack(message_id, C.FULFILLED, note=note)

    def auto_ack_rejected(self, message_id: str, reason: str = "", error_code: int = C.VALIDATION_ERROR) -> SendResult:
        """Send REJECTED ACK to indicate processing rejection."""
        return self.send_ack(message_id, C.REJECTED, error_code=error_code, note=reason)

    def auto_ack_failed(self, message_id: str, error: Exception) -> SendResult:
        """Send FAILED ACK to indicate processing failure."""
        error_code = self.error_mapper.map_exception(error)
        return self.send_ack(message_id, C.FAILED, error_code=error_code, note=str(error))


class MessageProcessor:
    """High-level message processor with automatic ACK handling."""

    def __init__(self, ack_manager: ACKLifecycleManager):
        self.ack_manager = ack_manager
        self._message_handlers: Dict[int, Callable[[Dict[str, Any]], Any]] = {}
        self._default_handler: Optional[Callable[[Dict[str, Any]], Any]] = None

    def register_handler(self, message_type: int, handler: Callable[[Dict[str, Any]], Any]) -> None:
        """Register a handler for a specific message type."""
        self._message_handlers[message_type] = handler

    def set_default_handler(self, handler: Callable[[Dict[str, Any]], Any]) -> None:
        """Set a default handler for unregistered message types."""
        self._default_handler = handler

    def process_message(self, envelope: Dict[str, Any]) -> SendResult:
        """Process an incoming message with automatic ACK handling."""
        message_type = envelope.get("message_type", C.MESSAGE_TYPE_UNSPECIFIED)
        message_id = envelope.get("message_id", "")

        try:
            # Send RECEIVED ACK immediately
            self.ack_manager.auto_ack_received(envelope)

            # Handle ACK messages specially
            if message_type == C.ACKNOWLEDGEMENT:
                self.ack_manager.process_incoming_ack(envelope)
                return SendResult(success=True, message_id=message_id, accepted=True, reason="ACK processed")

            # Send READ ACK before processing
            self.ack_manager.auto_ack_read(message_id)

            # Find and execute handler
            handler = self._message_handlers.get(message_type, self._default_handler)
            
            if handler:
                result = handler(envelope)
                # Send FULFILLED ACK on successful processing
                return self.ack_manager.auto_ack_fulfilled(message_id, f"Processed by {handler.__name__}")
            else:
                # No handler available - reject
                return self.ack_manager.auto_ack_rejected(
                    message_id, 
                    f"No handler for message type {message_type}",
                    C.UNSUPPORTED_MESSAGE_TYPE
                )

        except Exception as e:
            # Send FAILED ACK on processing error
            return self.ack_manager.auto_ack_failed(message_id, e)
